// @mui-treasury/component-flex/Flex.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Flex = exports.FlexCol = exports.FlexRow = exports.Item = exports.flexClasses = void 0;

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _unstyled = require("@material-ui/unstyled");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var flexClasses = (0, _unstyled.generateUtilityClasses)("Flex", ["row", "col", "item"]);
exports.flexClasses = flexClasses;

var Item = function Item(_ref) {
  var className = _ref.className,
      noShrink = _ref.noShrink,
      grow = _ref.grow,
      props = _objectWithoutProperties(_ref, ["className", "noShrink", "grow"]);

  return /*#__PURE__*/_react["default"].createElement(_Box["default"], _extends({
    className: (0, _clsx["default"])(flexClasses.item, className),
    minWidth: "0px"
  }, noShrink && {
    flexShrink: 0
  }, props, grow && {
    flexGrow: grow === "stretched" ? 100000 : 1
  }));
};

exports.Item = Item;

var resolveGap = function resolveGap(gap) {
  return (0, _utils.checkFlexGap)() ? {
    gap: gap
  } : {
    m: (0, _utils.halveGap)(gap, true),
    "& > *, & > .MuiBox-root": {
      p: (0, _utils.halveGap)(gap)
    }
  };
};

var FlexRow = function FlexRow(_ref2) {
  var className = _ref2.className,
      gap = _ref2.gap,
      noWrap = _ref2.noWrap,
      stackPoint = _ref2.stackPoint,
      props = _objectWithoutProperties(_ref2, ["className", "gap", "noWrap", "stackPoint"]);

  var resolvedGap = resolveGap(gap);
  return /*#__PURE__*/_react["default"].createElement(Item, _extends({
    className: (0, _clsx["default"])(flexClasses.row, className),
    minWidth: "0px"
  }, props, {
    sx: _extends({}, props.sx, {
      display: "flex",
      flexDirection: "row",
      flexWrap: noWrap ? "nowrap" : "wrap"
    }, resolvedGap, {
      "& > *, & > .MuiBox-root": _extends({}, resolvedGap["& > *, & > .MuiBox-root"], stackPoint && {
        minWidth: "clamp(0px, (".concat(typeof stackPoint === "number" ? "".concat(stackPoint, "px") : stackPoint, " + 1px - 100%) * 999, 100%)")
      })
    })
  }));
};

exports.FlexRow = FlexRow;

var FlexCol = function FlexCol(_ref3) {
  var className = _ref3.className,
      gap = _ref3.gap,
      props = _objectWithoutProperties(_ref3, ["className", "gap"]);

  return /*#__PURE__*/_react["default"].createElement(Item, _extends({
    className: (0, _clsx["default"])(flexClasses.col, className)
  }, props, {
    sx: _extends({}, props.sx, {
      display: "flex",
      flexDirection: "column"
    }, resolveGap(gap))
  }));
};

exports.FlexCol = FlexCol;
var Flex = {
  Row: FlexRow,
  Col: FlexCol,
  Item: Item
};
exports.Flex = Flex;
