// react-vis-styles/dist/index.esm.js
import React, { Fragment } from 'react';
import { Global, css } from '@emotion/core';

var LegendsStyle = function LegendsStyle(_ref) {
  var enabledColor = _ref.enabledColor,
      disabledColor = _ref.disabledColor;
  return React.createElement(Global, {
    styles:
    /*#__PURE__*/
    css(".rv-discrete-color-legend{box-sizing:border-box;overflow-y:auto;font-size:12px;&.horizontal{white-space:nowrap;}}.rv-discrete-color-legend-item{color:", enabledColor, ";border-radius:1px;padding:9px 10px;&.horizontal{display:inline-block;.rv-discrete-color-legend-item__title{margin-left:0;display:block;}}}.rv-discrete-color-legend-item__color{display:inline-block;vertical-align:middle;overflow:visible;}.rv-discrete-color-legend-item__color__path{stroke:#dcdcdc;stroke-width:2px;}.rv-discrete-color-legend-item__title{margin-left:10px;}.rv-discrete-color-legend-item.disabled{color:", disabledColor, ";}.rv-discrete-color-legend-item.clickable{cursor:pointer;&:hover{background:#f9f9f9;}}.rv-search-wrapper{display:flex;flex-direction:column;}.rv-search-wrapper__form{flex:0;}.rv-search-wrapper__form__input{width:100%;color:#a6a6a5;border:1px solid #e5e5e4;padding:7px 10px;font-size:12px;box-sizing:border-box;border-radius:2px;margin:0 0 9px;outline:0;}.rv-search-wrapper__contents{flex:1;overflow:auto;}.rv-continuous-color-legend{font-size:12px;.rv-gradient{height:4px;border-radius:2px;margin-bottom:5px;}}.rv-continuous-size-legend{font-size:12px;.rv-bubbles{text-align:justify;overflow:hidden;margin-bottom:5px;width:100%;}.rv-bubble{background:#d8d9dc;display:inline-block;vertical-align:bottom;}.rv-spacer{display:inline-block;font-size:0;line-height:0;width:100%;}}.rv-legend-titles{height:16px;position:relative;}.rv-legend-titles__left,.rv-legend-titles__right,.rv-legend-titles__center{position:absolute;white-space:nowrap;overflow:hidden;}.rv-legend-titles__center{display:block;text-align:center;width:100%;}.rv-legend-titles__right{right:0;}")
  });
};

LegendsStyle.defaultProps = {
  enabledColor: '#3a3a48',
  disabledColor: '#b8b8b8'
};

var PlotStyle = function PlotStyle(_ref) {
  var axisFontColor = _ref.axisFontColor,
      axisLineColor = _ref.axisLineColor,
      axisFontSize = _ref.axisFontSize,
      tooltipBackground = _ref.tooltipBackground,
      tooltipColor = _ref.tooltipColor,
      tooltipFontSize = _ref.tooltipFontSize,
      tooltipBorderRadius = _ref.tooltipBorderRadius,
      tooltipShadow = _ref.tooltipShadow,
      tooltipPadding = _ref.tooltipPadding;
  return React.createElement(Global, {
    styles:
    /*#__PURE__*/
    css(".rv-xy-plot{color:#c3c3c3;position:relative;canvas{pointer-events:none;}.rv-xy-canvas{pointer-events:none;position:absolute;}}.rv-xy-plot__inner{display:block;}.rv-xy-plot__axis__line{fill:none;stroke-width:2px;stroke:", axisLineColor, ";}.rv-xy-plot__axis__tick__line{stroke:", axisLineColor, ";}.rv-xy-plot__axis__tick__text{fill:", axisFontColor, ";font-size:", axisFontSize, ";}.rv-xy-plot__axis__title{text{fill:", axisFontColor, ";font-size:", axisFontSize, ";}}.rv-xy-plot__grid-lines__line{stroke:", axisLineColor, ";}.rv-xy-plot__circular-grid-lines__line{fill-opacity:0;stroke:", axisLineColor, ";}.rv-xy-plot__series,.rv-xy-plot__series path{pointer-events:all;}.rv-xy-plot__series--line{fill:none;stroke:#000;stroke-width:2px;}.rv-crosshair{position:absolute;font-size:11px;pointer-events:none;}.rv-crosshair__line{background:#47d3d9;width:1px;}.rv-crosshair__inner{position:absolute;text-align:left;top:0;}.rv-crosshair__inner__content{border-radius:", tooltipBorderRadius, ";background:", tooltipBackground, ";color:", tooltipColor, ";font-size:", tooltipFontSize, ";padding:", tooltipPadding, ";box-shadow:", tooltipShadow, ";}.rv-crosshair__inner--left{right:4px;}.rv-crosshair__inner--right{left:4px;}.rv-crosshair__title{font-weight:bold;white-space:nowrap;}.rv-crosshair__item{white-space:nowrap;}.rv-hint{position:absolute;pointer-events:none;}.rv-hint__content{border-radius:", tooltipBorderRadius, ";padding:", tooltipPadding, ";font-size:", tooltipFontSize, ";background:", tooltipBackground, ";box-shadow:", tooltipShadow, ";color:", tooltipColor, ";text-align:left;white-space:nowrap;}")
  });
};

PlotStyle.defaultProps = {
  axisFontColor: '#6b6b76',
  axisLineColor: '#e6e6e9',
  axisFontSize: '11px',
  tooltipBackground: '#3a3a48',
  tooltipColor: '#fff',
  tooltipFontSize: '12px',
  tooltipBorderRadius: '4px',
  tooltipShadow: '0 2px 4px rgba(0, 0, 0, 0.5)',
  tooltipPadding: '7px 10px'
};

var _ref = {
  name: "yzwoux",
  styles: ".rv-radial-chart{.rv-xy-plot__series--label{pointer-events:none;}}"
};

var RadialChartStyle = function RadialChartStyle() {
  return React.createElement(Global, {
    styles: _ref
  });
};

var _ref$1 = {
  name: "gzd8m0",
  styles: ".rv-treemap{font-size:12px;position:relative;}.rv-treemap__leaf{overflow:hidden;position:absolute;}.rv-treemap__leaf--circle{align-items:center;border-radius:100%;display:flex;justify-content:center;}.rv-treemap__leaf__content{overflow:hidden;padding:10px;text-overflow:ellipsis;}"
};

var TreemapStyle = function TreemapStyle() {
  return React.createElement(Global, {
    styles: _ref$1
  });
};

var RVStyle = function RVStyle() {
  return React.createElement(Fragment, null, React.createElement(LegendsStyle, null), React.createElement(PlotStyle, null), React.createElement(RadialChartStyle, null), React.createElement(TreemapStyle, null));
};

export default RVStyle;
export { LegendsStyle, PlotStyle, RadialChartStyle, TreemapStyle };
